File: 01. Arrays/08. Majority Element I/main.js
// Brute Force
function majorityElementBruteForce(nums) {
    for (let num of nums) {
        let count = nums.filter(x => x === num).length;
        if (count > nums.length / 2) return num;
    }
    return -1;
}

// Sorting
function majorityElementSorting(nums) {
    nums.sort((a, b) => a - b);
    return nums[Math.floor(nums.length / 2)];
}

// HashMap
function majorityElementHashMap(nums) {
    let freq = {};
    for (let num of nums) freq[num] = (freq[num] || 0) + 1;
    for (let key in freq) {
        if (freq[key] > nums.length / 2) return parseInt(key);
    }
    return -1;
}

// Boyer-Moore Voting Algorithm (Optimal)
function majorityElementBoyerMoore(nums) {
    let candidate = 0, count = 0;
    for (let num of nums) {
        if (count === 0) candidate = num;
        count += (num === candidate) ? 1 : -1;
    }
    return candidate;
}

// Main Function
let nums = [3, 2, 3];
console.log("Brute Force:", majorityElementBruteForce(nums));
console.log("Sorting:", majorityElementSorting(nums));
console.log("HashMap:", majorityElementHashMap(nums));
console.log("Boyer-Moore:", majorityElementBoyerMoore(nums));

File: 01. Arrays/08. Majority Element I/main.py
from collections import Counter

# Brute Force
def majority_element_brute_force(nums):
    for num in nums:
        if nums.count(num) > len(nums) // 2:
            return num
    return -1

# Sorting
def majority_element_sorting(nums):
    nums.sort()
    return nums[len(nums) // 2]

# HashMap
def majority_element_hashmap(nums):
    count = Counter(nums)
    return max(count.keys(), key=lambda k: count[k])

# Boyer-Moore Voting Algorithm (Optimal)
def majority_element_boyer_moore(nums):
    candidate, count = 0, 0
    for num in nums:
        if count == 0:
            candidate = num
        count += 1 if num == candidate else -1
    return candidate

# Main Function
nums = [3, 2, 3]
print("Brute Force:", majority_element_brute_force(nums))
print("Sorting:", majority_element_sorting(nums))
print("HashMap:", majority_element_hashmap(nums))
print("Boyer-Moore:", majority_element_boyer_moore(nums))
